### PR TITLE
Enhancement/skale 4744 change beneficiary

### PR DESCRIPTION
**Pull request by yavrsky**

### CODE REVIEW

File: contracts/Escrow.sol [hunk 1]
```diff
@@ -49,13 +49,19 @@ contract Escrow is IERC777Recipient, IERC777Sender, Permissions {
```
Comment (DimaStebaev) [inline contracts/Escrow.sol:50]: All `skale-manager` interfaces have to be imported from `skale-manager-interfaces` package.

```
import "./interfaces/delegation/IDelegationController.sol";
import "./interfaces/delegation/IDistributor.sol";
import "./interfaces/delegation/ITokenState.sol";
```

File: .solhint.json [hunk 2]
```diff
@@ -11,7 +11,7 @@
     "check-send-result": "error",
     "private-vars-leading-underscore": "error",
     "code-complexity": "error",
-    "func-order": "error",
+    "ordering": "error",
     "visibility-modifier-order": "error",
     "func-name-mixedcase": "error",
```
Comment (DimaStebaev) [inline .solhint.json:15]: Please set `comprehensive-interface` as an error and add all needed interfaces to `interfaces/` folder. Maybe we would publish them as a separate package in the future.

File: contracts/Allocator.sol [hunk 3]
```diff
@@ -36,6 +37,7 @@ import "./Permissions.sol";
  * @title Allocator
  */
 contract Allocator is Permissions, IERC777Recipient {
+    using ClonesUpgradeable for address;
```
Comment (DimaStebaev) [inline contracts/Allocator.sol]: Where is this library used?

File: contracts/Allocator.sol [hunk 4]
```diff
@@ -26,7 +26,8 @@ pragma experimental ABIEncoderV2;
 import "@openzeppelin/contracts-ethereum-package/contracts/introspection/IERC1820Registry.sol";
 import "@openzeppelin/contracts-ethereum-package/contracts/token/ERC777/IERC777Recipient.sol";
 import "@openzeppelin/contracts-ethereum-package/contracts/token/ERC20/IERC20.sol";
-import "./interfaces/openzeppelin/IProxyFactory.sol";
+import "@openzeppelin/contracts/proxy/TransparentUpgradeableProxy.sol";
+import "@openzeppelin/contracts-upgradeable/proxy/ClonesUpgradeable.sol";
 import "./interfaces/openzeppelin/IProxyAdmin.sol";
```
Comment (DimaStebaev) [inline contracts/Allocator.sol]: `IProxyAdmin` must be imported from `@openzeppelin/contracts-upgradeable` package.

File: contracts/Escrow.sol [hunk 5]
```diff
@@ -91,6 +101,19 @@ contract Escrow is IERC777Recipient, IERC777Sender, Permissions {
         _erc1820.setInterfaceImplementer(address(this), keccak256("ERC777TokensSender"), address(this));
     } 
 
+    function grantBeneficiaryRole(address account) external onlyBeneficiary {
```
Comment (DimaStebaev) [inline contracts/Escrow.sol]: Could you add a test of this function?
I suspect that `msg.sender` has to have `DEFAULT_ADMIN_ROLE` in the `Escrow` to be able to grant the role.

I recommend to remove this function and setup admin role for `BENEFICIARY_ROLE` and use standard function to grant it.

File: contracts/Escrow.sol [hunk 6]
```diff
@@ -91,6 +101,19 @@ contract Escrow is IERC777Recipient, IERC777Sender, Permissions {
...
+        super.grantRole(BENEFICIARY_ROLE, account);
+    }
+
+    function revokeBeneficiaryRole(address account) external onlyBeneficiary {
```
Comment (DimaStebaev) [inline contracts/Escrow.sol]: Could you add a test of this function?
I suspect that `msg.sender` has to have `DEFAULT_ADMIN_ROLE` in the `Escrow` to be able to revoke the role.

I recommend to remove this function and setup admin role for `BENEFICIARY_ROLE` and use standard function to revoke it.

File: contracts/Escrow.sol [hunk 7]
```diff
@@ -91,6 +101,19 @@ contract Escrow is IERC777Recipient, IERC777Sender, Permissions {
...
+        super.revokeRole(BENEFICIARY_ROLE, account);
+    }
+
+    function changeBeneficiary(address beneficiary) external onlyBeneficiary {
+        require(beneficiary != address(0), "Beneficiary address must not be zero");
+        _beneficiary = beneficiary;
```
Comment (DimaStebaev) [inline contracts/Escrow.sol:114]: There are a lot of mappings in the `Allocator` contract that use a beneficiary address. So changing it is more complex process.

File: contracts/Escrow.sol [hunk 8]
```diff
@@ -91,6 +101,19 @@ contract Escrow is IERC777Recipient, IERC777Sender, Permissions {
...
 
+    function grantBeneficiaryRole(address account) external onlyBeneficiary {
+        super.grantRole(BENEFICIARY_ROLE, account);
+    }
+
+    function revokeBeneficiaryRole(address account) external onlyBeneficiary {
+        super.revokeRole(BENEFICIARY_ROLE, account);
+    }
+
+    function changeBeneficiary(address beneficiary) external onlyBeneficiary {
```
Comment (DimaStebaev) [inline contracts/Escrow.sol]: What do you think about an idea to split this function to pair of `approve`/`change` ones? It ensures that beneficiary properly set new address and has ability to send transactions from it.

File: contracts/Permissions.sol [hunk 9]
```diff
@@ -31,9 +32,9 @@ import "./interfaces/IContractManager.sol";
...
  * @author Artem Payvin
  */
-contract Permissions is AccessControlUpgradeSafe {
-    using SafeMath for uint;
-    using Address for address;
+contract Permissions is InitializableWithGap, AccessControlUpgradeable {
+    using SafeMathUpgradeable for uint;
+    using AddressUpgradeable for address;
 
     IContractManager public contractManager;
```
Comment (DimaStebaev) [inline contracts/Permissions.sol:37]: Let's deprecate using of `skale-manager`'s `ContractManager`. There are just 2 smart contract. `ContractManager` is redundant.
This improves 2 things:
- reduces gas usage
- removes dependency on `skale-manager` that simplifies deployment and testing

`Permissions` itself also can be deprecated.

File: contracts/test/ProxyFactoryMock.sol [hunk 10]
```diff
@@ -63,20 +63,20 @@ contract ProxyMock {
         }
```
Comment (DimaStebaev) [inline contracts/test/ProxyFactoryMock.sol]: Can this file be removed?
Is it used anywhere?

File: contracts/test/SafeMock.sol [hunk 11]
```diff
@@ -0,0 +1,211 @@
+// SPDX-License-Identifier: AGPL-3.0-only
```
Comment (DimaStebaev) [inline contracts/test/SafeMock.sol:1]: This contract is included into `@skalenetwork/upgrade-tools`.

File: hardhat.config.ts [hunk 12]
```diff
@@ -0,0 +1,96 @@
...
+    return "auto";
+  }
+}
+
+const config: HardhatUserConfig = {
+  defaultNetwork: "hardhat",
+  solidity: {
+    compilers: [
+      {
+        version: '0.6.10',
```
Comment (DimaStebaev) [inline hardhat.config.ts]: Update solidity to modern version.

File: contracts/test/TimeHelpersTester.sol [hunk 13]
```diff
@@ -42,6 +42,15 @@ contract TimeHelpersTester is ITimeHelpers {
         return timestampToMonth(now);
     }
 
+    function monthToTimestamp(uint256 month) public view override returns (uint256 timestamp) {
+        uint256 year = _ZERO_YEAR;
+        uint256 _month = month;
+        year = year.add(_month.div(12));
```
Comment (DimaStebaev) [inline contracts/test/TimeHelpersTester.sol]: Solidity 8 has safe math included.

File: hardhat.config.ts [hunk 14]
```diff
@@ -0,0 +1,96 @@
...
+import "@nomiclabs/hardhat-etherscan";
+import "solidity-coverage";
+import "hardhat-typechain";
+import * as dotenv from "dotenv"
+import { utils, Wallet } from "ethers";
+import { HardhatNetworkAccountUserConfig } from "hardhat/types/config";
+
+dotenv.config();
+
+task("erc1820", "Deploy ERC1820 contract", async (_, { ethers }) => {
```
Comment (DimaStebaev) [inline hardhat.config.ts]: Do we need this task here?

File: migrations/deploy.ts [hunk 15]
```diff
@@ -0,0 +1,110 @@
...
+}
+
+export const contracts = [
+    "Allocator",
+    "Escrow"
+]
+
+async function main() {
+    if (await ethers.provider.getCode("0x1820a4B7618BdE71Dce8cdc73aAB6C95905faD24") === "0x") {
+        await run("erc1820");
```
Comment (DimaStebaev) [inline migrations/deploy.ts]: Do we need `ERC1820` in the allocator?

File: migrations/deploy.ts [hunk 16]
```diff
@@ -0,0 +1,110 @@
...
+    }
+
+    if (!existsSync(__dirname + "/../scripts/manager.json")) {
```
Comment (DimaStebaev) [inline migrations/deploy.ts:31]: If we deprecate `ContractManager` we can avoid this dependency.

File: migrations/deploy.ts [hunk 17]
```diff
@@ -0,0 +1,110 @@
...
+import { ethers, upgrades, network, run, artifacts } from "hardhat";
+import { getAbi } from './tools/abi';
+import { verifyProxy } from './tools/verification';
+import { Manifest } from "@openzeppelin/upgrades-core";
+import { Allocator } from '../typechain';
+import { getVersion } from './tools/version';
+import chalk from "chalk";
+
+
+export function getContractKeyInAbiFile(contract: string) {
```
Comment (DimaStebaev) [inline migrations/deploy.ts]: Many of functions below can be imported from the `@skalenetwork/upgrade-tools`.

File: migrations/tools/abi.ts [hunk 18]
```diff
@@ -0,0 +1,16 @@
+import { Interface } from "ethers/lib/utils";
```
Comment (DimaStebaev) [inline migrations/tools/abi.ts]: This file is not required if `@skalenetwork/upgrade-tools` is used.

File: migrations/tools/gnosis-safe.ts [hunk 19]
```diff
@@ -0,0 +1,233 @@
+import axios from "axios";
```
Comment (DimaStebaev) [inline migrations/tools/gnosis-safe.ts]: This file is not required if `@skalenetwork/upgrade-tools` is used.

File: migrations/tools/multiSend.ts [hunk 20]
```diff
@@ -0,0 +1,48 @@
+import { BigNumber } from "ethers";
```
Comment (DimaStebaev) [inline migrations/tools/multiSend.ts]: This file is not required if `@skalenetwork/upgrade-tools` is used.

File: migrations/tools/verification.ts [hunk 21]
```diff
@@ -0,0 +1,32 @@
+import { ethers, run, network } from "hardhat";
```
Comment (DimaStebaev) [inline migrations/tools/verification.ts]: This file is not required if `@skalenetwork/upgrade-tools` is used.

File: migrations/tools/version.ts [hunk 22]
```diff
@@ -0,0 +1,16 @@
+import { promises as fs } from 'fs';
```
Comment (DimaStebaev) [inline migrations/tools/version.ts]: This file is not required if `@skalenetwork/upgrade-tools` is used.

File: migrations/upgrade.ts [hunk 23]
```diff
@@ -0,0 +1,254 @@
...
+import { Allocator, SafeMock } from "../typechain";
+import { getVersion } from "./tools/version";
+import { getAbi } from "./tools/abi";
+import { verify } from "./tools/verification";
+import { exec as asyncExec } from "child_process";
+import util from 'util';
+const exec = util.promisify(asyncExec);
+
+
+async function upgrade(targetVersion: string, contractNamesToUpgrade: string[]) {
```
Comment (DimaStebaev) [inline migrations/upgrade.ts]: I suggest to import this function from the `@skalenetwork/upgrade-tools`.

File: migrations/upgrade.ts [hunk 24]
```diff
@@ -0,0 +1,254 @@
...
+
+    await fs.writeFile(`data/skale-allocator-${version}-${network.name}-abi.json`, JSON.stringify(abi, null, 4));
+
+    console.log("Done");
+}
+
+async function main() {
+    await upgrade(
+        "2.2.0",
+        contracts
```
Comment (DimaStebaev) [inline migrations/upgrade.ts]: We need to add custom initialization of 3 `Escrow`'s on mainnet.

File: openzeppelin-cli-export.json [hunk 25]
```diff
@@ -0,0 +1,44 @@
+{
```
Comment (DimaStebaev) [inline openzeppelin-cli-export.json]: I guess this file can be removed.

File: package.json [hunk 26]
```diff
@@ -22,57 +22,59 @@
...
+    "compile": "npx hardhat clean && npx hardhat compile",
     "hooks": "git config core.hooksPath .githooks || true",
+    "no-hooks": "git config core.hooksPath .git/hooks",
     "lint": "npx solhint \"contracts/**/*.sol\"",
-    "prepare": "yarn generate && yarn hooks",
-    "test": "npx buidler test && yarn tsc",
+    "prepare": "yarn compile && yarn hooks",
+    "test": "yarn tsc && npx hardhat test",
     "tsc": "tsc --noEmit",
     "tslint": "npx tslint --project .",
```
Comment (DimaStebaev) [inline package.json]: `tslint` has been deprecated. I must be replaced with `eslint`.

File: package.json [hunk 27]
```diff
@@ -22,57 +22,59 @@
...
     "slither": "slither .",
     "fullcheck": "yarn lint && yarn tsc && yarn tslint && yarn slither"
   },
   "dependencies": {
     "@openzeppelin/cli": "^2.8.2",
+    "@openzeppelin/contracts": "3.4.0",
```
Comment (DimaStebaev) [inline package.json]: Let's remove with dependency to not be confused with upgradeable version.

File: package.json [hunk 28]
```diff
@@ -22,57 +22,59 @@
...
-    "test": "npx buidler test && yarn tsc",
+    "prepare": "yarn compile && yarn hooks",
+    "test": "yarn tsc && npx hardhat test",
     "tsc": "tsc --noEmit",
     "tslint": "npx tslint --project .",
     "slither": "slither .",
     "fullcheck": "yarn lint && yarn tsc && yarn tslint && yarn slither"
   },
   "dependencies": {
     "@openzeppelin/cli": "^2.8.2",
```
Comment (DimaStebaev) [inline package.json]: openzeppelin cli has been deprecated.

**Review by payvint — DISMISSED** [review]
Legend

File: contracts/Allocator.sol [hunk 29]
```diff
@@ -112,13 +72,42 @@ contract Allocator is Permissions, IERC777Recipient {
...
-        external override
+        external
+        override
         allow("SkaleToken")
         // solhint-disable-next-line no-empty-blocks
     {
 
     }
 
+    function requestBeneficiaryAddress(address newBeneficiaryAddress) external override {
```
Comment (DimaStebaev) [inline contracts/Allocator.sol]: Maybe `requestBeneficiaryAddressChange` or `changeBeneficiaryAddress`?

File: contracts/Allocator.sol [hunk 30]
```diff
@@ -556,10 +537,7 @@ contract Allocator is Permissions, IERC777Recipient {
      *     if current step is 17 and vesting interval is 7 function returns 21.
      */
     function _calculateNextVestingStep(uint256 currentStep, uint256 vestingInterval) private pure returns (uint256) {
-        return currentStep
-            .add(vestingInterval)
-            .sub(
-                currentStep.mod(vestingInterval)
-            );
+        // slither-disable-next-line weak-prng
```
Comment (DimaStebaev) [inline contracts/Allocator.sol:540]: It this relevant?

File: contracts/Escrow.sol [hunk 31]
```diff
@@ -41,25 +43,23 @@ import "./Permissions.sol";
  * @title Escrow
  * @dev This contract manages funds locked by the Allocator contract.
  */
-contract Escrow is IERC777Recipient, IERC777Sender, Permissions {
+contract Escrow is IERC777Recipient, IERC777Sender, IEscrow, Permissions {
 
     address internal _beneficiary;
 
+    // slither-disable-next-line unused-state
     uint256 private _availableAmountAfterTermination;
```
Comment (DimaStebaev) [inline contracts/Escrow.sol:49]: Why is it unused?

File: contracts/Escrow.sol [hunk 32]
```diff
@@ -41,25 +43,23 @@ import "./Permissions.sol";
...
 
     IERC1820Registry private _erc1820;
 
-    bytes32 public constant BENEFICIARY_ROLE = keccak256("BENEFICIARY_ROLE");
```
Comment (DimaStebaev) [inline contracts/Escrow.sol]: Why did you remove this?

File: contracts/Permissions.sol [hunk 33]
```diff
@@ -19,23 +19,22 @@
...
 /**
  * @title Permissions - connected module for Upgradeable approach, knows ContractManager
  * @author Artem Payvin
  */
-contract Permissions is InitializableWithGap, AccessControlUpgradeable {
-    using SafeMathUpgradeable for uint;
+contract Permissions is AccessControlUpgradeableLegacy, IPermissions {
     using AddressUpgradeable for address;
 
+    // slither-disable-next-line uninitialized-state
```
Comment (DimaStebaev) [inline contracts/Permissions.sol]: Delete this value somewhere in the code. It frees the state and remove unuzed-variable warning.

File: contracts/interfaces/delegation/IDelegationController.sol [hunk 34]
```diff
@@ -19,7 +19,7 @@
     along with SKALE Allocator.  If not, see <https://www.gnu.org/licenses/>.
```
Comment (DimaStebaev) [inline contracts/interfaces/delegation/IDelegationController.sol:20]: This file must be deleted in favor of `@skalenetwork/skale-manager-interfaces`.

File: scripts/update_manifest.ts [hunk 35]
```diff
@@ -2,6 +2,7 @@ import { exec as asyncExec } from "child_process";
 import { ethers, upgrades } from "hardhat";
```
Comment (DimaStebaev) [inline scripts/update_manifest.ts:2]: Do we still need this script? Maybe it can be deleted?

**Review by payvint — DISMISSED** [review]

**Review by payvint — DISMISSED** [review]
Legend again!

**Review by payvint — DISMISSED** [review]

File: migrations/upgrade.ts [hunk 36]
```diff
@@ -0,0 +1,279 @@
...
+
+type CustomEscrowsAction = (proxyAdmin: ProxyAdmin, proxies: string[], newImplementationAddress: string, safeTransactions: string[]) => Promise<void>;
+interface CustomEscrows {
+    [escrow: string]: {
+        oldImplementation: string
+        beneficiary: string
+    }
+}
+
+async function upgrade(targetVersion: string, contractNamesToUpgrade: string[], upgradeCustomEscrows: CustomEscrowsAction) {
```
Comment (DimaStebaev) [inline migrations/upgrade.ts]: Can the `upgrade` function from `upgrade-tools` be used?

File: migrations/upgrade.ts [hunk 37]
```diff
@@ -0,0 +1,279 @@
...
+        contracts,
+        async(proxyAdmin, proxies, newImplementationAddress, safeTransactions) => {
+            const escrowFactory = await ethers.getContractFactory("Escrow");
+            const escrows = JSON.parse(await fs.readFile(__dirname + "/../data/customEscrows.json", "utf-8")) as CustomEscrows;
+            for (const escrow in escrows) {
+                if (escrows[escrow].oldImplementation == await proxyAdmin.getProxyImplementation(escrow)) {
+                    if (escrows[escrow].beneficiary == "") {
+                        throw Error("Beneficiary wasn't found");
+                    }
+                    const encodedReinitialize = escrowFactory.interface.encodeFunctionData("reinitialize", [escrows[escrow].beneficiary]);
```
Comment (DimaStebaev) [inline migrations/upgrade.ts]: Are you sure that a beneficiary must be passed here but not an admin?

File: migrations/upgrade.ts [hunk 38]
```diff
@@ -0,0 +1,279 @@
...
+                    safeTransactions.push(encodeTransaction(
+                        0,
+                        proxyAdmin.address,
+                        0,
+                        proxyAdmin.interface.encodeFunctionData("upgradeAndCall", [escrow, newImplementationAddress, encodedReinitialize])
+                    ));
+                }
+                const index = proxies.indexOf(escrow);
+                if (~index) {
```
Comment (DimaStebaev) [inline migrations/upgrade.ts]: Why `~` but not `!`?

File: scripts/test_deploy.sh [hunk 39]
```diff
@@ -2,18 +2,28 @@
...
 cp data/skale-manager-*-abi.json ../scripts/manager.json
 cd ..
 rm -r --interactive=never skale-manager
 
-NODE_OPTIONS="--max-old-space-size=4096" PRODUCTION=true npx truffle migrate --network test
-sleep 5
+nvm use $CURRENT_NODE_VERSION
+
+NODE_OPTIONS="--max-old-space-size=4096" npx hardhat run migrations/deploy.ts --network localhost
 kill $GANACHE_PID
```
Comment (DimaStebaev) [inline scripts/test_deploy.sh]: This does not work in nodejs 16. I suggest to replace with `kill-port`.

**Review by DimaStebaev — APPROVED** [review]
